ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/barrier_receiver.cpp

```diff
@@ -0,0 +1,121 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ethernet/dataflow_api.h"
+#include "tt_metal/hw/inc/ethernet/dataflow_api.h"
+#include <array>
+#include "ttnn/cpp/ttnn/operations/ccl/kernels/edm/edm_handshake.hpp"
+#define MIN_WAIT 100000
+
+FORCE_INLINE void perform_rs_loop(
+    uint64_t channel_sem_addr,
+    volatile eth_channel_sync_t* eth_channel_syncs
+    )
+{
+    //In the Ring Start case, we first send the signal to the sender and then we wait on the ethernet
+    noc_semaphore_inc(channel_sem_addr, 1);
+    uint32_t count = 0;
+    while(eth_channel_syncs->bytes_sent == 0) {
+        if (count == MIN_WAIT)
+        {
+            count = 0;
+            run_routing();
+        }
+        else
+        {
+            asm volatile ("");
+            count ++;
+        }
+    }
+    eth_channel_syncs->bytes_sent = 0;
+    while (eth_txq_is_busy());
```

Comment:
This will need a `run_routing`escape path. I think it makes sense to add a `wait_for_eth_txq_cmd_space(wait_min)`